core/cmd/app_test.go

```diff
@@ -48,6 +48,116 @@ var (
 			},
 		},
 	}
+
+	allow                           = false
+	dbURL                           = "postgres://chainlink:mysecretpassword@172.17.0.1:5432/primary"
+	backupDbURL                     = "postgres://chainlink:mysecretpassword@172.17.0.1:5433/replica"
+	testSecretsFileContentsComplete = chainlink.Secrets{
+		Secrets: v2.Secrets{
+			Database: v2.DatabaseSecrets{
+				URL:                  models.NewSecretURL(models.MustParseURL(dbURL)),
+				BackupURL:            models.NewSecretURL(models.MustParseURL(backupDbURL)),
+				AllowSimplePasswords: &allow,
+			},
+			Explorer: v2.ExplorerSecrets{
+				AccessKey: models.NewSecret("EXPLORER_ACCESS_KEY"),
+				Secret:    models.NewSecret("EXPLORER_TOKEN"),
+			},
+			Password: v2.Passwords{
+				Keystore: models.NewSecret("mysecretpassword"),
+				VRF:      models.NewSecret("mysecretvrfpassword"),
+			},
+			Pyroscope: v2.PyroscopeSecrets{
+				AuthToken: models.NewSecret("PYROSCOPE_TOKEN"),
+			},
+			Prometheus: v2.PrometheusSecrets{
+				AuthToken: models.NewSecret("PROM_TOKEN"),
+			},
+			Mercury: v2.MercurySecrets{
+				Credentials: map[string]v2.MercuryCredentials{
+					"key1": {
+						URL:      models.NewSecretURL(models.MustParseURL("https://mercury.stage.link")),
+						Username: models.NewSecret("user"),
+						Password: models.NewSecret("user_pass"),
+					},
+					"key2": {
+						URL:      models.NewSecretURL(models.MustParseURL("https://mercury.stage.link")),
+						Username: models.NewSecret("user"),
+						Password: models.NewSecret("user_pass"),
+					},
+				},
+			},
+			Threshold: v2.ThresholdKeyShareSecrets{
+				ThresholdKeyShare: models.NewSecret("THRESHOLD_SECRET"),
+			},
+		},
+	}
+
+	testSecretsRedactedContentsComplete = chainlink.Secrets{
+		Secrets: v2.Secrets{
+			Database: v2.DatabaseSecrets{
+				URL:                  models.NewSecretURL(models.MustParseURL("xxxxx")),
+				BackupURL:            models.NewSecretURL(models.MustParseURL("xxxxx")),
```

Comment:
Why does explicit `xxxxx` need to be used here? I'm worried that could mask problems with the printing protection, since it matches w/o conversion and could lead to a false positive. Comparing either the raw TOML strings, or Go types with real values would be cleaner, if possible.

Reply:
Hmm, yea - good point!

IMO, three things should be tested here:
1. Merge is successful for all non-overriding operations
2. Merge fails in presence of any overriding operations
3. Verify merge results in the correct ending value of secrets

The first two test cases are addressed sufficiently through the current changes.  The third test case is not. The crux is that when the new merging is performed, `New()`,  is also functionally coupled with the redaction of the secrets. The redaction logic is defined in `core/store/models/secrets.go`. As an aside, I believe it makes sense that tests 1 and 2 are performed in `app_test.go` since end to end error and success conditions for use of the node are tested here. 

I've split the merging and redaction functionality in the next commit to isolate and unit test the merging functionality in `config_general_test.go`. However - given how the secret types are implemented, there is no direct way to compare by types or by toml string. I'd either have to create a new type for testing that is similar to the Secrets type, or implement a new toml string outputter that won't redact the secret values. The latter approach seems like it could be potentially abused.  The solution I’ve landed on is comparing the values in `opts` before `New()` is called and the values behind the `Secret` and `SecretURL` types in opts after New() is called through the use of a new `TestSecrets` type.

Reply:
I'm not sure that I follow. The redaction logic prevents printing/marshaling directly, but the values are still directly available and comparable. If these tests _depend_ on the literal value being `xxxxx`, then something seems wrong. If they don't depend on _any_ particular values, then that also seems wrong. I thought we would have real values here in the Go types, and a raw TOML string of redacted types elsewhere. Are we reading in a redacted file somewhere or something like that?

Reply:
No, like you mentioned on a comment below - by design it's not possible to go from the secret types to non-redacted values in the TOML. I'll use raw TOML as you suggested, it should simplify all of the ...workarounds I used for this revision to test that third case 😆